Sources/Machine/rbm_spin_v2.cc

```diff
@@ -0,0 +1,325 @@
+  }
+  if (b_.has_value()) {
+    parameters.segment(i, b_->size()) = *b_;
+    i += b_->size();
+  }
+  parameters.segment(i, W_.size()) =
+      Eigen::Map<Eigen::VectorXcd>(W_.data(), W_.size());
+  return parameters;
+}
+
+void RbmSpinV2::SetParameters(Eigen::Ref<const Eigen::VectorXcd> parameters) {
+  if (parameters.size() != Npar()) {
+    std::ostringstream msg;
+    msg << "wrong shape: [" << parameters.size() << "]; expected [" << Npar()
+        << "]";
+    throw InvalidInputError{msg.str()};
+  }
+  Index i = 0;
+  if (a_.has_value()) {
+    *a_ = parameters.segment(i, a_->size());
+    i += a_->size();
+  }
+  if (b_.has_value()) {
+    *b_ = parameters.segment(i, b_->size());
+    i += b_->size();
+  }
+  Eigen::Map<Eigen::VectorXcd>(W_.data(), W_.size()) =
+      parameters.segment(i, W_.size());
+}
+
+void RbmSpinV2::LogVal(Eigen::Ref<const RowMatrix<double>> x,
+                       Eigen::Ref<Eigen::VectorXcd> out, const any &) {
+  if (x.cols() != Nvisible()) {
+    std::ostringstream msg;
+    msg << "input tensor has wrong shape: [" << x.rows() << ", " << x.cols()
+        << "]; expected [?"
+        << ", " << Nvisible() << "]";
+    throw InvalidInputError{msg.str()};
+  }
+  if (out.size() != x.rows()) {
+    std::ostringstream msg;
+    msg << "output tensor wrong shape: [" << out.size() << "]; expected ["
+        << x.rows() << "]";
+    throw InvalidInputError{msg.str()};
+  }
+  BatchSize(x.rows());
+  if (a_.has_value()) {
+    out.noalias() = x * (*a_);
+  } else {
+    out.setZero();
+  }
+  theta_.noalias() = x * W_;
+  ApplyBiasAndActivation(out);
+}
+
+void RbmSpinV2::DerLog(Eigen::Ref<const RowMatrix<double>> x,
+                       Eigen::Ref<RowMatrix<Complex>> out,
+                       const any & /*unused*/) {
+  if (x.cols() != Nvisible()) {
+    std::ostringstream msg;
+    msg << "input tensor has wrong shape: [" << x.rows() << ", " << x.cols()
+        << "]; expected [?"
+        << ", " << Nvisible() << "]";
+    throw InvalidInputError{msg.str()};
+  }
+  if (out.rows() != x.rows() || out.cols() != Npar()) {
+    std::ostringstream msg;
+    msg << "output tensor wrong shape: [" << out.rows() << ", " << out.cols()
+        << "]; expected [" << x.rows() << ", " << Npar() << "]";
+    throw InvalidInputError{msg.str()};
+  }
+  BatchSize(x.rows());
+
+  auto i = Index{0};
+  if (a_.has_value()) {
+    out.block(0, i, BatchSize(), a_->size()) = x;
+    i += a_->size();
+  }
+
+  Eigen::Map<RowMatrix<Complex>>{theta_.data(), theta_.rows(), theta_.cols()}
+      .noalias() = x * W_;
+  if (b_.has_value()) {
+    theta_.array() = (theta_ + b_->transpose().colwise().replicate(BatchSize()))
+                         .array()
+                         .tanh();
+    out.block(0, i, BatchSize(), b_->size()) = theta_;
+    i += b_->size();
+  } else {
+    theta_.array() = theta_.array().tanh();
+  }
+
+  // TODO: Rewrite this using tensors
+#pragma omp parallel for schedule(static)
```

Comment:
maybe I'd remove explicit calls to omp, this might conflict with MPI / negatively affect performance on multiple cores?